### PR TITLE
Reinstate working sessions in browser 

### PR DESCRIPTION
**Pull request by rsafonseca**
I've seen that session persistence stopped working due to https://github.com/apache/cloudstack/commit/19e3c0168e744a76b5e1dc24a5eafa776d342404
From what I could gather from the comments, this was done to fix issue where separate instances of management servers could exist within the same domain.
I've fixed the above mentioned issue by prepending the location's hostname, in order to allow both sessions to co-exist without clashing.
This also removes the need for this fix https://github.com/apache/cloudstack/commit/6c71d3bae1a3a72a9fa4004decdba4a7174f6913

### CODE REVIEW

**Comment by rsafonseca** [comment]
@bhaisaab please take a look :)

File: ui/scripts/cloudStack.js [hunk 1]
```diff
@@ -329,7 +325,6 @@
                         g_regionsecondaryenabled = null;
                         g_loginCmdText = null;
                         
-                        $.cookie('JSESSIONID', null);
                         $.cookie('sessionkey', null);
```
Comment (terbolous) [inline ui/scripts/cloudStack.js]: I'm no UI/Javascript expert, but should these include hostname as well? If so, same goes for L351+

Comment (rsafonseca) [inline ui/scripts/cloudStack.js]: Indeed Erik, thanks for pointing that out. I've updated the PR.

**Comment by rsafonseca** [comment]
I searched around for that reference and found nothing useful, perhaps @bhaisaab can shed some light, since it was signed off by him :)
If I understood correctly, the issue is easy to reproduce, two instances of management server in the same domain would get messed up with each other's auth due to the cookies.
This prevents said issue from occurring, hopefully Rohit will confirm that :)

File: ui/scripts/cloudStack.js [hunk 2]
```diff
@@ -117,22 +117,15 @@
                         }
                         return cookieValue;
                     };
-                    unBoxCookieValue('sessionkey');
-                    // if sessionkey cookie exists use this to set g_sessionKey
```
Comment (bhaisaab) [inline ui/scripts/cloudStack.js]: The original issue was related to security where session key was stored in cookie; the crux of the fix by Jessica was to use the cookie to set the sessionkey on window (g_sessionKey variable) and destroy it so it may not be read later.

**Comment by vadimkim** [comment]
Have tested with 4.5.1 -- works good !

**Comment by rsafonseca** [comment]
I see that in the parent commit Jessica tried to fix security on the cookie by setting httponly=true in tomcat's context .xml
This would be a sane thing to do, if the sessionkey cookie was actually handled by tomcat, which is not.
This cookie only existed on client side, and was created by .js, so it's not possible to set httponly on that.
This approach does not add any extra security, since the data is still stored in a .js variable and accessible to scripts, only the location of the data has changed, but this not confer extra security.

Basically, all this did was break the functionality and remove 1 of the 2 places where the sessionkey was accessible through javascript.

My new proposal is to actually get httponly cookie working, which thwarts most kinds of XSS attacks.
This is done by making the server handle the cookie (not the client).

The only thing i didn't fix was the same security hole being created by the SAML login (cookie with sessionkey data is created in browser), since i didn't have it setup to test. These changes shouldn't break it though, it will just keep the same insecure behaviour.

Have a look at my proposed changes :)

**Comment by stephen-turner** [comment]
As this changes behaviour which was aiming to fix a security problem, I've referred it to the security mailing list.

**Comment by bhaisaab** [comment]
Hi @rsafonseca thanks for the update and patch. It's a good idea to use HTTPOnly flag on cookie to ensure we have the authentication going on and at the same time not expose the cookie value to javascript. But the current patch will break SAML and any redirected authentication which relies on cookies being read by JS and then destroyed. As an alternative, let's discuss how we can improve this:

1. When a user is authenticated, we set the sessionkey cookie (httponly). This btw, only protects it from being read by JavaScript and not MITM attacks. For this, we can use a global setting I had added (isSecureSessionCookieEnabled method in ApiServer can be used for this) to enable secure flag on it as well. Please use Cookie class for this instead of adding raw strings like: resp.addCookie(new cookie(with flags etc enabled.)
2. In HttpUtils, add helper method:
+    public static String findCookie(final Cookie[] cookies, String key) {
+        for (Cookie cookie: cookies) {
+            if (cookie.getName().equals(key)) {
+                return cookie.getValue();
+            }
+        }
+        return null;
+    }
3. In ApiServlet, use the above method to find the cookie value (if available) when there is no sessionkey or empty sessionkey passed in the API params, check against session's sessionkey value to invalidate session. 
4. When users use default login method (login command), the login response still has the sessionkey. So, we can discuss to remove this as a security measure as well. If users are using an external logging system such as SAML based, usually the plugin would set the cookie. If the cookie becomes httponly (so JavaScript won't be able to read it), we will need to fix the JavaScript code such that it does not rely on sessionkey parameter at all.

Please review the above ideas and comment/share on them.

TLDR? To fix the login/UI issue, I think we need to not depend on passing a sessionkey as API argument but instead use HTTPOnly (and enable secure flag, so cookies are passed on HTTPs if enabled) "sessionkey" cookie along with the JSESSIONID cookie. Doing this would make sure we can run ACS UI across tabs and let the UI work when a external auth system such as those based on SAML, 2-factor or OAuth can work.

**Comment by rsafonseca** [comment]
Hi Rohit, thank you for your view on this :)
Here's my comment on it:

1. No use on protecting the sessionkey from MITM attacks, as the username/password are sent in plain text, and so are all the other commands, so anyone can pickup your credentials if they're listening on the wire ;). For MITM protection SSL should be used. I did notice that isSecureSessionCookieEnabled i actually placed the cookie in raw to be exactly the same as the sessioncookie. I did use the cookie class for reading though :). So.. both cookies should be inserted in the same fashion.

2. Will take care of that :)

3. I miss the purpose.. session cleanup? Can you explain a bit more? :)

4. If you don't return it in the loginResponse, regular (non-browser) calls to the API will fail, because i guess all integrations with the API depend on it, that's why i didn't remove it. It can be in the response, as long as javascript doesn't store it for later access.

Regarding the non-bulleted stuff:

Secure cookie:
I don't agree with this one, that will make it just not work at all in HTTP. We aim to fix a browser vulnerability here, not a communication channel one. If someone wants on the wire security, they should just switch to SSL.

Not depend on sessionkey as API argument:
Will that not break some API implementations? I think supporting both methods is the best for now.

About SAML and redirect based auth not working:
Are you sure? Did you test? I left it in a way it would likely work... cookie is still read an put into the javascript variable if it was created on redirection. Since the SAML plugin just puts it back in a cookie so i didn't bother destroying it in that scenario, the plugin should be modded for that. Either way.. it *should* work :)

Thanks for reviewing.

**Comment by bhaisaab** [comment]
Hi @rsafonseca, thanks for replying.

- I agree, it's best to force and use SSL than use secure flag.
- Regarding point 3, what I mean to explain is that since rely on sessionkey passed in the HTTP API parameter when we set sessionkey httponly cookie we should put a fix to check sessionkey in the cookie as well on a request. The issue here is that for normal login it would work, because the login response will send the sessionkey which javascript code will save in a g_ (global) variable and reuse it for future API requests. In case of alternative auth/login, for example the SAML plugin, we rely only on cookies but if you make the sessionkey cookie httponly - the JS code won't be able to read that and future APIs will send empty sessionkey API arg/parameter. Now, to make this work - the APIServlet needs to check for sessionkey cookie in case sessionkey HTTP parameter is missing or empty. The SAML plugin would set a sessionkey cookie that is readable but since in the fix after we authenticate() you're setting sessionkey cookie to be httponly (so after login/redirection, JS code won't be able to read that cookie). If this is confusing, leave it for me - I can help fix this.
- Regarding point 4, I guess what I meant to say is not to remove the argument but cover the above case. I agree we don't need to remove it since API clients may break.

I guess we have some agreement here, please go ahead and make the changes. I'm working on making the SAML plugin production grade, so I can help test all aspects of logging/authentication around it as well.

Cheers.

**Comment by rsafonseca** [comment]
Hi @bhaisaab 
I don't think that is an issue :)
Whatever your login method, the cookie will always be set on the first successful command and subsequent ones.
On normal login, the js doesn't really store it in a variable, it tries to read the httponly cookie and fails , and all subsequent API commands have the param sessionkey="null".. that's why i'm checking if command.equals("null") to get the cookie :)
If you have tested and have issues, might also be the case that the saml codes tries to set the cookie, overwriting the existing httponly cookie with a plain one, thus breaking auth.

**Comment by rsafonseca** [comment]
Hi @bhaisaab 

Any news on testing it? If i misunderstood the problem you stated, can you explain further or help fix as you suggested?
I'm sure a lot of people would like to see this through :)

**Comment by bhaisaab** [comment]
@rsafonseca yeah, can you fix the cookie usage? Sorry due to dayjob projects I'm not getting time to review PRs and test them actively. I'm planning to test and merge this next week along with some of the SAML improvements I'm doing, since merging that work may conflict (or maybe not) with the auth subsystem.

Meanwhile, if possible please see other ways of improving it - for example, since you're not habituated like rest of us -- you can better see the auth subsystem, security and UI from your fresh eyes and perhaps think of how we can improve it. For example, why do we have JSESSIONID when we actually rely on sessionkey, those kinds of things.

**Comment by DaanHoogland** [comment]
This merges cleanly, should we wait on extra fixes? or have them done in separate PRs?

**Comment by vadimkim** [comment]
I have recently implemented SSL proxy in front of CS. This way seems to be more practical than deploying certificates directly into Tomcat.  After such modification session persistence does not work anymore. So basically I see no point of this change anymore. May be we need to find better solution if possible?

**Comment by rsafonseca** [comment]
Session persistence is not working on any scenario atm. The aim of this is to fix that. The security measure is aimed to prevent browser vulnerability, which using SSL doesn't prevent. Why do you say this is not needed any more?

**Comment by vadimkim** [comment]
I have applied you patch manually and it worked for HTTP. I have expected that if I put HTTPS proxy in front of it  - it will continue to work. But at HTTPS session if I do "F5" to refresh page -- I am logged off. So from the end-user perspective there is no (visual) difference. I would like to understand what problem are we going to solve? It seems for me it is not clearly defined. Or may be I understood it in a wrong way. I have expected that F5 does not brake session.

**Comment by rsafonseca** [comment]
@vadimkim 
Let me try to break it down:

In the past, with the purpose of preventing a browser vulnerability, session persistence was completely broken, so currently, there is no session persistence with either HTTP or HTTPS from ACS 4.3.x to 4.6.x, that is, if you press F5, you will be logged out.

This fix reenables persistence, and does so while improving on the security measure that was the objective that led to session persistence being broken.

With this patch, both HTTP and HTTPS are working fine with session persistence, and the browser vulnerability (XSS attack) is eliminated. SSL would not protect against a XSS attack, it would only protect from the fact that you are sending your credentials and sessionkey unencrypted from your machine to the server. If you are doing so over a secure network that you trust, using SSL might even be irrelevant, depending on the situation. So, SSL will only protect your credentials/session on the transport layer, other nasty XSS scripts might still gain access to whatever is stored in your browser. Setting the cookie as HttpOnly prevents your own browser from reading it, and thus makes that information safe from XSS.

If you're having issues with a reverse proxy in front of it, the issue is probably with the reverse proxy or browser config... here's where i see it might be going wrong on your setup:
- Proxy is destroying the cookie;
- Proxy is not allowing the cookie in the first place;
- Proxy if forcing a new session to be generated;
- Browser isn't allowing cookies on the https url (perhaps configured manually to always reject cookies and an exception for the http:// cloudstack url ?)
- Proxy might be delivering a cached version of cloudStack.js without the fix;

Could you share some more info on your reverse proxy setup? I'll can try to reproduce your issue and give you a hand :)

**Comment by vadimkim** [comment]
@rsafonseca 
Thank you for explanation. I will try to put my points here:
1. For public cloud HTTPS must be used (this is not an option) and external SSL offload (proxy) should not brake entire logic.
2. While I have found, that most of the logic works just perfect behind the proxy there are 2 issues that need to be fixed 
* login procedure somewhere uses direct path (not relative) and tried to call /client/api?command directly. I have fixed it by 1 rewrite rule (see below). Otherwise login is not possible -- 404 response.
* F5 brakes session if I work over HTTPS proxy. Hope we will fix it together.

I am using very simple nginx conf. Pure proxy:
```
server {
    listen 443 ssl;
    server_name kickcloud.net;
    ssl_certificate           /usr/local/etc/ssl/my.crt;
    ssl_certificate_key       /usr/local/etc/ssl/my.key;

    ssl_session_cache  builtin:1000  shared:SSL:10m;
    ssl_protocols  TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers HIGH:!aNULL:!eNULL:!EXPORT:!CAMELLIA:!DES:!MD5:!PSK:!RC4;
    ssl_prefer_server_ciphers on;

    access_log            /var/log/nginx/kickckloud.net.log;

    location / {
      rewrite /client(.*) $1  last;
      proxy_set_header        X-Real-IP $remote_addr;
      proxy_set_header        X-Forwarded-Server $host;
      proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header        X-Forwarded-Proto $scheme;
      proxy_pass          http://x.x.x.x:8080/client/;
      proxy_redirect      default;
      proxy_read_timeout  90;
    }
  }
```

x.x.x.x -- backend server where Cloudstack runs
public name that you can access  and check *cloudStack.js* is here: https://kickcloud.net/. I have some more redirects from ports 8080 and 80 to 443, but they are not relevant here I assume.

**Comment by rsafonseca** [comment]
@vadimkim 
It seems like you're not passing to cookies over to the client on that config :)

Try setting:
    proxy_pass_header  Set-Cookie;
    proxy_pass_header  P3P;

Since your internal domain may not be the same as the external, you might also need some additional config like:
    proxy_cookie_domain     kickcloud.net     $host;
    sub_filter      "kickcloud.net"   $host;
Or something similar :)

I will try to reproduce your setup when i have some time to figure out a config that works for sure, but try these configs in the mean while :)

Either way, this seems like an issue with the proxy config itself, and there should be no need to adjust code for that situation ;)

**Comment by rsafonseca** [comment]
About the redirect rule, i will soon bake a patch that will render that unnecessary, but for now, there should be no problem with just rewriting it in the proxy :)
I'll discuss this with the others on wednesday's meetup in SBP

**Comment by vadimkim** [comment]
@rsafonseca 
I have added Set-Cookie and P3P lines into config. My $host variable contains either kickcloud.net or www.kickcloud.net -- depends on request user made. So I didn't set cookie_domain and sub_filter, because domain is the same.  With such a setup I am still logged off after F5.  There is something else that we are missing.

**Comment by rsafonseca** [comment]
@vadimkim 
I need to do some testing with this setup, which i'm afraid i might only have time to do later this week or in the weekend. It generally easier to do this with httpd (using ProxyPassReverseCookieDomain  and ProxyPassReverseCookiePath), but i'll figure out a way to make it work on nginx :)
The thing is, your internal domain might not be the same as your external domain, and the same goes for the cookie path. You will most likely need to rewrite those bits in the cookie eg. cloudstack.myintra.local <-> kickcloud.net or something like that.
I'll get back to you as soon as i can test it and hopefully provide you a working config :)

**Comment by vadimkim** [comment]
@rsafonseca 
Can you help me to define the correct internal domain name?  If I check cookie at plain HTTP server (where it works) I see **domainid** field with UUID. At HTTPS server I see the same parameter with the same UUID. For me both cookies looks the same except cookie name. Should it be name of the cookie that must be rewritten?

**Comment by rsafonseca** [comment]
@vadimkim 
The only cookie that is set by the server and needs to be rewritten is the sessionkey cookie, all the others are created by the client and will point to the domain and path in the end user's perspective. This is the cookie you need to check and rewrite.

**Comment by vadimkim** [comment]
@rsafonseca 
I didn't get it. Don't have so  deep knowledge about cookie manipulation.  If I open dev-tools at Chrome  I can see only 1 cookie with name = hostname where I connect to and fields that also have names like $host_userid, $host_sessionKey, etc.  I shouldn't replace anything here because it is client cookie, isn't it ?  But where do I get server cookie then?

**Comment by rsafonseca** [comment]
@vadimkim are you building the latest version of this PR? I've removed the initial host_ prefix on the cookies, since that was not the issue that led to breaking the sessions.
Commit https://github.com/rsafonseca/cloudstack/commit/7c2502a9c2cd538336c5d87a7613f3b541bd70ec
Removes that suffix and makes the cookie server side
If you're seing a host_ prefix on the cookies, you shouldn't have a server set cookie either

**Comment by vadimkim** [comment]
@rsafonseca 
I didn't apply this patch indeed. It requires JAVA class re-compilation and I don't dare to make it directly at production. JS was not so critical from that point.
Development environment is not yet set-up on my site, so I will live without session persistence for some time. Thank you for detailed explanation. Now I understand better what is going on during log-in and what is required to make session persistent. Will be waiting for these changes at mainstream.
